tianshou/env/vecenv/shmemenv.py

```diff
@@ -0,0 +1,159 @@
+import ctypes
+from collections import OrderedDict
+from multiprocessing import Pipe, Process, Array
+from typing import Callable, List, Optional, Tuple, Union
+
+import gym
+import numpy as np
+
+from tianshou.env import BaseVectorEnv, SubprocVectorEnv
+from tianshou.env.utils import CloudpickleWrapper
+
+_NP_TO_CT = {np.float64: ctypes.c_double,
+             np.float32: ctypes.c_float,
+             np.int32: ctypes.c_int32,
+             np.int8: ctypes.c_int8,
+             np.uint8: ctypes.c_char,
+             np.bool: ctypes.c_bool}
+
+
+def _shmem_worker(parent, p, env_fn_wrapper, obs_bufs,
+                  obs_shapes, obs_dtypes, keys):
+    """Control a single environment instance using IPC and
+    shared memory.
+    """
+    def _encode_obs(maybe_dict):
+        flatdict = maybe_dict if isinstance(maybe_dict, dict) else {
+            None: maybe_dict}
+        for k in keys:
+            dst = obs_bufs[k].get_obj()
+            dst_np = np.frombuffer(dst, dtype=obs_dtypes[k]).reshape(
+                obs_shapes[k])
+            np.copyto(dst_np, flatdict[k])
+        return None
+
+    parent.close()
+    env = env_fn_wrapper.data()
+    try:
+        while True:
+            cmd, data = p.recv()
+            if cmd == 'step':
+                obs, reward, done, info = env.step(data)
+                p.send((_encode_obs(obs), reward, done, info))
+            elif cmd == 'reset':
+                p.send(_encode_obs(env.reset()))
+            elif cmd == 'close':
+                p.send(env.close())
+                p.close()
+                break
+            elif cmd == 'render':
+                p.send(env.render(**data) if hasattr(env, 'render') else None)
+            elif cmd == 'seed':
+                p.send(env.seed(data) if hasattr(env, 'seed') else None)
+            elif cmd == 'getattr':
+                p.send(getattr(env, data) if hasattr(env, data) else None)
+            else:
+                p.close()
+                raise NotImplementedError
+    except KeyboardInterrupt:
+        p.close()
+
+
+class ShmemVectorEnv(SubprocVectorEnv):
+    """Optimized version of SubprocVectorEnv that uses shared
+    variables to communicate observations.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.env.BaseVectorEnv` for more detailed
+        explanation.
+    """
+    """I borrowed heavily from openai baseline to implement ShmemVectorEnv Class.
```

Comment:
No need to again `"""`, merge them together.

Reply:
fixed